### PR TITLE
FEATURE: Add Neos.CliSetup package

### PR DESCRIPTION
**Pull request by mficzel**
The package contains the commands

`./flow welcome` - Show a welcome message with informations how to setup neos via cli. This should be used later as post-create-project-cmd for the neos-distributions.
`./flow setup:database` - Configure the database connection interactively
`./flow setup:imagehandler` - Configure the image handling interactively

The solution is implemented as a separate package in the development collection for the following reasons:
- The browser based setup tools `Neos.Setup` + `Neos.NeosSetup` may get deprecated one day
- The commands are not added to an existing package as they should be uninstallable
- The development should happen in the dev-collection to be able to evolve with Neos

Resolves: #3506

Todos after merging:

- [ ] configure package subsplit (jenkins)
- [ ] register package via packagist
- [ ] add package to distributions (require + post-create-project-script)

### CODE REVIEW

**Comment by mficzel** [comment]
It would be great if someone would look over the texts during the review as i tend to sneak in typos and weird formulations.

**Comment by mficzel** [comment]
Notes: 
- select best available image driver 
- store persistence and media config in separate files.

File: Neos.CliSetup/Classes/Command/SetupCommandController.php [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+                sprintf('DB Driver (<info>%s</info>): ', $this->persistenceConfiguration['driver']),
+                $availableDrivers,
+                $this->persistenceConfiguration['driver']
+            );
+        }
+
+        if (is_null($host)) {
+            $host = $this->output->ask(
+                sprintf('Host (<info>%s</info>): ', $this->persistenceConfiguration['host'] ?? '127.0.0.1'),
+                $this->persistenceConfiguration['host'] ?? '127.0.0.1'
```
Comment (mhsdesign) [inline Neos.CliSetup/Classes/Command/SetupCommandController.php:88]: just wondering is ?? the right one or is '' returned?

**Review by mhsdesign — COMMENTED** [review]
quick note...

File: Neos.CliSetup/Classes/Command/SetupCommandController.php (hunk 1 above)
Comment (mficzel) [reply Neos.CliSetup/Classes/Command/SetupCommandController.php:88]: This falls back to '127.0.0.1' if $this->persistenceConfiguration['host'] is null or does not exist.
See: https://www.php.net/manual/en/migration70.new-features.php#migration70.new-features.null-coalesce-op

Tested it again and it works as expected but i noticed that other empty values show empty brackets which is confusion so i added three dashes there `---`. Nut entirely sure wether `null` would be better understandable.

File: Neos.CliSetup/Classes/Command/SetupCommandController.php [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+            $this->outputLine($e->getMessage());
+            $this->quit(1);
+        }
+
+        $filename = $this->getSettingsFilename('Settings.Database.yaml');
+
+        $this->outputLine();
+        $this->output(sprintf('<info>%s</info>',$this->writeSettings($filename, 'Neos.Flow.persistence.backendOptions',$persistenceConfiguration)));
+        $this->outputLine();
+        $this->outputLine(sprintf("The new database settings were written to file <info>%s</info>", $filename));
```
Comment (sorenmalling) [inline Neos.CliSetup/Classes/Command/SetupCommandController.php:133]: Suggestion: Remove "file" so the sentence goes

The new database settings were written to <info>%s</info>

File: Neos.CliSetup/Classes/Command/SetupCommandController.php [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+    /**
+     * @param string|null $driver
+     */
+    public function imageHandlerCommand(string $driver = null): void
+    {
+        $availableImageHandlers = $this->imageHandlerService->getAvailableImageHandlers();
+
+        if (count($availableImageHandlers) == 0) {
+            $this->outputLine("No supported image handler was found.");
```
Comment (sorenmalling) [inline Neos.CliSetup/Classes/Command/SetupCommandController.php:144]: Suggestion: Either remove "was" here or add it to "No supported database driver .." text above

File: Neos.CliSetup/Classes/Command/SetupCommandController.php [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+                $availableImageHandlers,
+                array_key_last($availableImageHandlers)
+            );
+        }
+
+        $filename = $this->getSettingsFilename('Settings.Imagehandling.yaml');
+        $this->outputLine();
+        $this->output(sprintf('<info>%s</info>', $this->writeSettings($filename, 'Neos.Imagine.driver', $driver)));
+        $this->outputLine();
+        $this->outputLine(sprintf("The new image handler setting were written to file <info>%s</info>", $filename));
```
Comment (sorenmalling) [inline Neos.CliSetup/Classes/Command/SetupCommandController.php:160]: Suggestion: Remove "file" here, similar to above

File: Neos.CliSetup/Classes/Command/WelcomeCommandController.php [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+                ....#####......############
+                ....#####  ......##########
+                ....#####    ......########
+                ....#####      ......######
+                .#######         ........
+
+            Welcome to Neos.
+            </info>
+
+            The following steps will help you to configure your Neos:
```
Comment (sorenmalling) [inline Neos.CliSetup/Classes/Command/WelcomeCommandController.php:46]: Suggestion Remove "Your" so the sentence goes

`The following steps will help you to configure Neos`

File: Neos.CliSetup/Classes/Command/WelcomeCommandController.php [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+            1. Configure the database connection:
+               <info>./flow setup:database</info>
+            2. Create the required database tables:
+               <info>./flow doctrine:migrate</info>
+            3. Configure the imageHandling:
```
Comment (sorenmalling) [inline Neos.CliSetup/Classes/Command/WelcomeCommandController.php:52]: Suggestion: Change `imageHandling` to `image handler` - the wording `imageHandling` is not used anywhere else

File: Neos.CliSetup/Classes/Command/SetupCommandController.php (hunk 2 above)
Comment (mficzel) [reply Neos.CliSetup/Classes/Command/SetupCommandController.php]: Applied as suggested thanks

File: Neos.CliSetup/Classes/Command/WelcomeCommandController.php (hunk 5 above)
Comment (mficzel) [reply Neos.CliSetup/Classes/Command/WelcomeCommandController.php]: Applied as suggested thanks

File: Neos.CliSetup/Classes/Command/WelcomeCommandController.php (hunk 6 above)
Comment (mficzel) [reply Neos.CliSetup/Classes/Command/WelcomeCommandController.php]: Applied as suggested thanks

File: Neos.CliSetup/Classes/Command/SetupCommandController.php (hunk 4 above)
Comment (mficzel) [reply Neos.CliSetup/Classes/Command/SetupCommandController.php]: Applied as suggested thanks

**Comment by lorenzulrich** [comment]
> It would be great if someone would look over the texts during the review as i tend to sneak in typos and weird formulations.

Texts looking good.

File: Neos.CliSetup/Classes/Command/SetupCommandController.php (hunk 1 above)
Comment (mhsdesign) [reply Neos.CliSetup/Classes/Command/SetupCommandController.php:88]: nice works well ;)

three dashes are better than () i guess. But null would be fine to...

**Comment by mhsdesign** [comment]
just another note: the current web setup ignores the context and writes in the non context specific folder/file. But i guess your behaviour is okay since for production its fine to reconfigure stuff?
(maybe it should be noted in the welcome screen that you need to use FLOW_CONTEXT=Production ./flow ... to use this command in production)

That the changes were made to context x *should* be clear by the file name info `The new image handler setting were written to Configuration/Development/Settings.Imagehandling.yaml`

last thing: i'd maybe like to now as a user if a database was created or just used. - so if thats easy to implement ...

**Comment by markusguenther** [comment]
Just looked into the code and did not test it yet. The code looks fine, will try to test it soonish so that we have a chance to get that into the release :)

Thanks, Martin for bringing that up

**Comment by mficzel** [comment]
Adjusted the following:
- Configuration is written to Configuration/Settings.Database.yaml and Configuration/Settings.Imagehanding.yaml instead of contexts as the application context is an advanced concept that can be a hurdle when trying Neos
- The setup:database command now shows informations about wether the database could be connected or a new one was created. Some parts of the logic from the DatabaseConnectionService were therefore moved to the commandController.

**Review by mhsdesign — APPROVED** [review]
Nice works all well. Thanks for the adjustments ;)

(except that you can provide nothing as database name in the dialogue and it says 'Database {insert null here} was sucessfully created.' - but i would take care of this and a more advanced and forgiving database connection/creation step.)

**Comment by markusguenther** [comment]
I finally managed to test the PR, and it is really cool to have that. As I mentioned before, code wise it looks good.
I now stumbled upon an issue with Postgres. I tried it with version 14 and 14.1 and when I use the cli command the new Database cannot be used, caused by encoding issues.

When I use the setup web tool, everything is fine.

Screenshot_2021-11-21_at_20_12_15

So with PSQL it is from my point of view not useable because the Database is the essential part ;)
But hope that it is easy to fix 🙈

Kudos @mficzel for all the work 👏

**Comment by markusguenther** [comment]
Ah with MySQL everything worked fine BTW ;)

**Comment by mhsdesign** [comment]
hmm the code used to actually create the db is excactly the same as in the web setup: 
(method createDatabase)
https://github.com/neos/setup/blob/2fd9b320ed455c0496a3b4f6db2b786baa9ffb4f/Classes/Step/DatabaseStep.php#L183

**Comment by mficzel** [comment]
@markusguenther will check postgres setup again but i also wonder what went wrong as the executed code is really identical.

**Comment by markusguenther** [comment]
I was also surprised because it worked in the browser :/

I use the brew installation on MacOS Monterey. Can also install a docker based DB and check it again.

**Review by sorenmalling — APPROVED** [review]

**Comment by markusguenther** [comment]
@sorenmalling did Postgres work for you?

**Comment by markusguenther** [comment]
Ok tried it again.

```
docker run --name postgresql-container -p 5432:5432 -e POSTGRES_PASSWORD=somePassword -d postgres
```

Then used the new neos commands.

```
./flow setup:database
```

And on the doctrine migrate it strikes.
I have MacOS with PHP 7.4.25 and the latest postgres container. 

Screenshot 2021-11-22 at 17 56 41


In the browser everything works 🤷‍♂️

**Comment by markusguenther** [comment]
I am fine with merging it, If we are confident to get this fixed. But I actually have no clue why that happens, TBH.

**Comment by mficzel** [comment]
@markusguenther i found it. postgres does not know about utf8mb4 and used utf8.

The old setup did this: https://github.com/neos/setup/blob/master/Classes/Step/DatabaseStep.php#L112-L114 

I added a charset `utf8` to the pg configuration and could create postgres dbs successfully.

**Comment by markusguenther** [comment]
Thanks @mficzel 💚

**Pull request merged**